From f612f9eeb806f1c7be4694d8ff49e01eea6e05a1 Mon Sep 17 00:00:00 2001
From: Lu Wei <luwei32@huawei.com>
Date: Thu, 13 Jan 2022 22:35:17 +0800
Subject: [PATCH 1005/2065] [Huawei] net: core: Add a GID field to struct sock.

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186640

--------------------------------

UID and GID are requested as filters for socketmap, but we can only get
UID from sock structure. This patch adds GID field to struct sock as UID.

Signed-off-by: Lu Wei <luwei32@huawei.com>
Signed-off-by: Liu Jian <liujian56@huawei.com>
---
 include/net/sock.h | 14 ++++++++++++++
 net/core/sock.c    |  2 ++
 net/socket.c       |  6 ++++--
 3 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/include/net/sock.h b/include/net/sock.h
index c958be11d172..af73dda0285b 100644
--- a/include/net/sock.h
+++ b/include/net/sock.h
@@ -303,6 +303,7 @@ struct bpf_local_storage;
   *	@sk_ack_backlog: current listen backlog
   *	@sk_max_ack_backlog: listen backlog set in listen()
   *	@sk_uid: user id of owner
+  *	@sk_gid: group id of owner
   *	@sk_priority: %SO_PRIORITY setting
   *	@sk_type: socket type (%SOCK_STREAM, etc)
   *	@sk_protocol: which protocol this socket belongs in this network family
@@ -527,7 +528,14 @@ struct sock {
 #endif
 	struct rcu_head		sk_rcu;
 
+#ifndef __GENKSYMS__
+	union {
+		kgid_t	sk_gid;
+		u64	sk_gid_padding;
+	};
+#else
 	KABI_RESERVE(1)
+#endif
 	KABI_RESERVE(2)
 	KABI_RESERVE(3)
 	KABI_RESERVE(4)
@@ -1904,6 +1912,7 @@ static inline void sock_graft(struct sock *sk, struct socket *parent)
 	parent->sk = sk;
 	sk_set_socket(sk, parent);
 	sk->sk_uid = SOCK_INODE(parent)->i_uid;
+	sk->sk_gid = SOCK_INODE(parent)->i_gid;
 	security_sock_graft(sk, parent);
 	write_unlock_bh(&sk->sk_callback_lock);
 }
@@ -1916,6 +1925,11 @@ static inline kuid_t sock_net_uid(const struct net *net, const struct sock *sk)
 	return sk ? sk->sk_uid : make_kuid(net->user_ns, 0);
 }
 
+static inline kgid_t sock_net_gid(const struct net *net, const struct sock *sk)
+{
+	return sk ? sk->sk_gid : make_kgid(net->user_ns, 0);
+}
+
 static inline u32 net_tx_rndhash(void)
 {
 	u32 v = prandom_u32();
diff --git a/net/core/sock.c b/net/core/sock.c
index bee3c320dbfe..2fa8863caee0 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -2985,9 +2985,11 @@ void sock_init_data(struct socket *sock, struct sock *sk)
 		RCU_INIT_POINTER(sk->sk_wq, &sock->wq);
 		sock->sk	=	sk;
 		sk->sk_uid	=	SOCK_INODE(sock)->i_uid;
+		sk->sk_gid	=	SOCK_INODE(sock)->i_gid;
 	} else {
 		RCU_INIT_POINTER(sk->sk_wq, NULL);
 		sk->sk_uid	=	make_kuid(sock_net(sk)->user_ns, 0);
+		sk->sk_gid	=	make_kgid(sock_net(sk)->user_ns, 0);
 	}
 
 	rwlock_init(&sk->sk_callback_lock);
diff --git a/net/socket.c b/net/socket.c
index d52c265ad449..7d84c289e5ae 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -543,10 +543,12 @@ static int sockfs_setattr(struct dentry *dentry, struct iattr *iattr)
 	if (!err && (iattr->ia_valid & ATTR_UID)) {
 		struct socket *sock = SOCKET_I(d_inode(dentry));
 
-		if (sock->sk)
+		if (sock->sk) {
 			sock->sk->sk_uid = iattr->ia_uid;
-		else
+			sock->sk->sk_gid = iattr->ia_gid;
+		} else {
 			err = -ENOENT;
+		}
 	}
 
 	return err;
-- 
2.17.1

